### PR TITLE
[REVIEW CAREFULLY] services are selective on which clusters they listen on

### PR DESCRIPTION
**Pull request by eishay**
up to now all services listened on all clusters. it loads on the ZK and the instances. 
having an explicit list will help us realize dependencies between services. 
by default, all services listen on their own cluster to know who the leader is.

@ymatsuda @raymondng @andrewconner @stkem @leogrim @yingjieMiao 
Please look at the `servicesToListenOn` list for each service and verify its not too long or short.

### CODE REVIEW

File: kifi-backend/modules/search/app/com/keepit/search/SearchProdModule.scala [hunk 1]
```diff
@@ -17,4 +19,9 @@ case class SearchProdModule() extends SearchModule(
   indexModule = ProdIndexModule(),
   trackingModule = ProdTrackingModule(),
   spellModule = SpellCorrectorModule()
-)with CommonProdModule
+) with CommonProdModule  {
+  val discoveryModule = new ProdDiscoveryModule {
+    def servicesToListenOn = ServiceType.SHOEBOX :: ServiceType.HEIMDAL :: Nil
```
Comment (ymatsuda) [inline kifi-backend/modules/search/app/com/keepit/search/SearchProdModule.scala]: I think Message  index needs Eliza.

File: kifi-backend/modules/heimdal/app/com/keepit/heimdal/HeimdalProdModule.scala [hunk 2]
```diff
@@ -2,8 +2,15 @@ package com.keepit.heimdal
...
+import com.keepit.common.zookeeper.ProdDiscoveryModule
+import com.keepit.common.service.ServiceType
 
 case class HeimdalProdModule() extends HeimdalModule(
   cacheModule = HeimdalCacheModule(MemcachedCacheModule(), EhCacheCacheModule()),
   mongoModule = ProdMongoModule()
-) with CommonProdModule
+) with CommonProdModule {
+  val discoveryModule = new ProdDiscoveryModule {
+    def servicesToListenOn = Nil //is that right?
```
Comment (leogrim) [inline kifi-backend/modules/heimdal/app/com/keepit/heimdal/HeimdalProdModule.scala]: Heimdal should listen on Shoebox

**Comment by leogrim** [comment]
This is what I gathered:
SearchServiceClient goes into Eliza, Shoebox
ShoeboxServiceClient goes into Eliza, Search, Heimdal, Scraper
HeimdalServiceClient goes into Eliza, Search, Shoebox
ScraperServiceClient goes into Shoebox
AbookServiceClient goes into Eliza, Shoebox
ElizaServiceClient goes into Search, Shoebox

File: kifi-backend/modules/eliza/app/com/keepit/eliza/ElizaProdModule.scala [hunk 3]
```diff
@@ -3,8 +3,15 @@ package com.keepit.eliza
...
+import com.keepit.common.service.ServiceType
 
 case class ElizaProdModule() extends ElizaModule(
   cacheModule = ElizaCacheModule(MemcachedCacheModule(), EhCacheCacheModule()),
   urbanAirshipModule = ElizaUrbanAirshipModule()
-) with CommonProdModule
+) with CommonProdModule {
+  val discoveryModule = new ProdDiscoveryModule {
+    def servicesToListenOn = ServiceType.SEARCH :: ServiceType.SHOEBOX :: ServiceType.HEIMDAL :: Nil
+  }
```
Comment (leogrim) [inline kifi-backend/modules/eliza/app/com/keepit/eliza/ElizaProdModule.scala:15]: Eliza also needs ABook

File: kifi-backend/modules/scraper/app/com/keepit/scraper/ProdScraperServiceModule.scala [hunk 4]
```diff
@@ -3,9 +3,16 @@ package com.keepit.scraper
...
+import com.keepit.common.service.ServiceType
 
 case class ProdScraperServiceModule() extends ScraperServiceModule(
   cacheModule = ScraperCacheModule(MemcachedCacheModule(), EhCacheCacheModule()),
   storeModule = ScraperProdStoreModule(),
   scrapeProcessorModule = ProdScraperProcessorModule()
-) with CommonProdModule
+) with CommonProdModule {
+  val discoveryModule = new ProdDiscoveryModule {
+    def servicesToListenOn = ServiceType.SHOEBOX :: ServiceType.HEIMDAL :: Nil
```
Comment (leogrim) [inline kifi-backend/modules/scraper/app/com/keepit/scraper/ProdScraperServiceModule.scala]: Scraper doesn't need Heimdal at this point

File: kifi-backend/modules/abook/app/com/keepit/abook/ABookProdModule.scala [hunk 5]
```diff
@@ -3,9 +3,15 @@ package com.keepit.abook
...
+import com.keepit.common.zookeeper.ProdDiscoveryModule
 
 case class ABookProdModule() extends ABookModule(
   cacheModule = ABookCacheModule(MemcachedCacheModule(), EhCacheCacheModule()),
   storeModule = ABookProdStoreModule(),
   contactsUpdaterPluginModule = ProdContactsUpdaterPluginModule()
-) with CommonProdModule
+) with CommonProdModule {
+  val discoveryModule = new ProdDiscoveryModule {
+    def servicesToListenOn = ServiceType.SHOEBOX :: ServiceType.HEIMDAL :: Nil
```
Comment (leogrim) [inline kifi-backend/modules/abook/app/com/keepit/abook/ABookProdModule.scala]: Looks like ABook doesn't need anything

Comment (raymondng) [inline kifi-backend/modules/abook/app/com/keepit/abook/ABookProdModule.scala]: Not today :)

**Comment by raymondng** [comment]
abook & scraper -- lgtm

**Comment by eishay** [comment]
Thanks!
I'll merge tomorrow morning after the Play2.2 upgrade will cool down

File: kifi-backend/modules/common/app/com/keepit/common/zookeeper/ServiceDiscovery.scala [hunk 6]
```diff
@@ -66,8 +65,11 @@ class ServiceDiscoveryImpl(
 
   private val clusters: TrieMap[ServiceType, ServiceCluster] = {
     val clustersToInit = new TrieMap[ServiceType, ServiceCluster]()
+    val myCluster = new ServiceCluster(services.currentService, airbrake, if (services.currentService == ServiceType.SHOEBOX) Some(scheduler) else None)
+    clustersToInit(services.currentService) = myCluster
+    if (servicesToListenOn.contains(services.currentService)) throw new IllegalArgumentException(s"current service is included in servicesToListenOn: $servicesToListenOn")
     servicesToListenOn foreach {service =>
-      val cluster = new ServiceCluster(service, airbrake, if (services.currentService==ServiceType.SHOEBOX) Some(scheduler) else None)
```
Comment (stkem) [inline kifi-backend/modules/common/app/com/keepit/common/zookeeper/ServiceDiscovery.scala:70]: I think this breaks the "cluster to small" alerting. Shoebox needs a scheduler in all service clusters to make the alerting work.

Comment (eishay) [inline kifi-backend/modules/common/app/com/keepit/common/zookeeper/ServiceDiscovery.scala:70]: fixed. now all services are checking for cluster size. if we have alerts then probably more then one service is going to scream but I guess its fine.
at least we won't have a dependency that no one checks.

**Pull request merged**